File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        //Monotonic stack (both increasing from left->right and right->left) for smallest element index
        stack<pair<int,int>> stackLeftToRight; 
        stack<pair<int,int>> stackRightToLeft; 
        int n = heights.size();
        vector<int> smallerElementIndexOnLeft(n,-1);
        vector<int> smallerElementIndexOnRight(n,n);
        for(int i=0;i<n;i++) {
            if(stackLeftToRight.empty()) {
                stackLeftToRight.push({heights[i],i});
            } else {
                while(stackLeftToRight.empty()==false && stackLeftToRight.top().first>=heights[i]) {
                    stackLeftToRight.pop();
                }
                if(stackLeftToRight.empty()) {
                    stackLeftToRight.push({heights[i],i});
                } else {
                    smallerElementIndexOnLeft[i]=stackLeftToRight.top().second;
                    stackLeftToRight.push({heights[i],i});


                }
            }
        }

        for(int i=n-1;i>=0;i--) {
            if(stackRightToLeft.empty()) {
                stackRightToLeft.push({heights[i],i});
            } else {
                while(stackRightToLeft.empty()==false && stackRightToLeft.top().first>=heights[i]) {
                    stackRightToLeft.pop();
                } 
               if(stackRightToLeft.empty()) {
                    stackRightToLeft.push({heights[i],i});
                } else {
                    smallerElementIndexOnRight[i]=stackRightToLeft.top().second;
                    stackRightToLeft.push({heights[i],i});


                }
            }
        }

    int ans=INT_MIN;
    for(int i=0;i<n;i++) {
    ans = max(ans,(smallerElementIndexOnRight[i]-smallerElementIndexOnLeft[i]-1)*heights[i]);   
        }
        return ans;
    }
};

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        nsr_stack,nsl_stack = [],[]
        n = len(heights)

        nsr,nsl = [n]*n,[-1]*n
        for i in range(n-1,-1,-1):
            while nsr_stack and heights[nsr_stack[-1]]>=heights[i]:
                nsr_stack.pop()
            if nsr_stack: nsr[i] = nsr_stack[-1]
            nsr_stack.append(i)
        for i in range(0,n,1):
            while nsl_stack and heights[nsl_stack[-1]]>=heights[i]:
                nsl_stack.pop()
            if nsl_stack: nsl[i] = nsl_stack[-1]
            nsl_stack.append(i)
        answer = 0
        for i in range(n):
            answer = max(answer,(nsr[i]-nsl[i]-1)*heights[i])
        return answer
        